### PR TITLE
Fast path

### PR DESCRIPTION
**Pull request by lemmy**
Related: https://github.com/grapheo12/pirateship/issues/3

### CODE REVIEW

File: pirateship.tla [hunk 1]
```diff
@@ -252,7 +264,11 @@ ReceiveEntries(r, p) ==
     \* the only time a crash commit index can decrease is on the receipt of a NewView message if there's been a byz attack
     /\ crashCommitIndex' = [crashCommitIndex EXCEPT ![r] = Max2(@, HighestCrashQC(log'[r]))]
     \* assumes that a replica can safely byz commit if there's a quorum certificate over a quorum certificate
-    /\ byzCommitIndex' = [byzCommitIndex EXCEPT ![r] = Max2(@, HighestQCOverQC(log'[r]))]
+    \* Compare: src/consensus/commit.rs#maybe_byzantine_commit
+    /\ LET bci == HighestQCOverQC(log'[r])
+           \* Compare: src/consensus/commit.rs#maybe_byzantine_commit_by_fast_path
+           bciFastPath == HighestUnanimity(log[p])
```
Comment (heidihoward) [inline pirateship.tla]:
```suggestion
           bciFastPath == HighestUnanimity(log'[p])
```

File: pirateship.tla [hunk 2]
```diff
@@ -320,9 +336,11 @@ ReceiveVote(p, r) ==
     /\ IF viewStable'[p] THEN 
             /\ crashCommitIndex' = [crashCommitIndex EXCEPT ![p] = 
                 MaxQuorum(CQ, log[p], prepareQC'[p], @)]
-            \* Compare: ssrc/consensus/commit.rs#maybe_byzantine_commit
-            /\ byzCommitIndex' = [byzCommitIndex EXCEPT ![p] = 
-                HighestByzQC(SubSeq(log[p],1,MaxQuorum(BQ, log[p], prepareQC'[p], 0)))]
+            \* Compare: src/consensus/commit.rs#maybe_byzantine_commit
+            /\ LET bci == HighestByzQC(SubSeq(log[p], 1, MaxQuorum(BQ, log[p], prepareQC'[p], 0)))
+                   \* Compare: src/consensus/commit.rs#maybe_byzantine_commit_by_fast_path
+                   bciFastPath == HighestUnanimity(log[p])
```
Comment (heidihoward) [inline pirateship.tla]: bcifastpath needs to use `prepareQC` to take the newly received vote into account

File: pirateship.tla [hunk 3]
```diff
@@ -252,7 +264,11 @@ ReceiveEntries(r, p) ==
     \* the only time a crash commit index can decrease is on the receipt of a NewView message if there's been a byz attack
     /\ crashCommitIndex' = [crashCommitIndex EXCEPT ![r] = Max2(@, HighestCrashQC(log'[r]))]
     \* assumes that a replica can safely byz commit if there's a quorum certificate over a quorum certificate
-    /\ byzCommitIndex' = [byzCommitIndex EXCEPT ![r] = Max2(@, HighestQCOverQC(log'[r]))]
+    \* Compare: src/consensus/commit.rs#maybe_byzantine_commit
+    /\ LET bci == HighestQCOverQC(log'[r])
+           \* Compare: src/consensus/commit.rs#maybe_byzantine_commit_by_fast_path
+           bciFastPath == HighestUnanimity(log'[p])
```
Comment (heidihoward) [inline pirateship.tla]:
```suggestion
           bciFastPath == HighestUnanimity(log'[r])
```

Comment (lemmy) [reply pirateship.tla]: https://github.com/heidihoward/pirateship-tla/pull/22/commits/0b6e92722ed0063be6e6650b006622a0650d2b2e

File: pirateship.tla (hunk 2 above)
Comment (lemmy) [reply pirateship.tla]: https://github.com/heidihoward/pirateship-tla/pull/22/commits/a60e37257148835ab873a9982af197a9fbdc6104

File: pirateship.tla [hunk 4]
```diff
@@ -138,6 +139,8 @@ TypeOK ==
     /\ viewStable \in [R -> BOOLEAN]
     /\ view \in [R -> Views]
     /\ log \in [R -> Log]
+    /\ \A l \in Range(log):
+        \A i \in DOMAIN l: l[i].byzQC = {} => l[i].byzQCVotes = {}
```
Comment (heidihoward) [inline pirateship.tla]:
```suggestion
        \A i \in DOMAIN l: l[i].byzQC = {} <=> l[i].byzQCVotes = {}
```

File: pirateship.tla [hunk 5]
```diff
@@ -189,6 +192,16 @@ HighestQCOverQC(l) ==
         idx == SelectLastInSubSeq(l, 1, lidx, IsByzQC)
     IN IF idx = 0 THEN 0 ELSE Max(l[idx].byzQC)
 
+HighestUnanimity(l, idx, r) ==
```
Comment (heidihoward) [inline pirateship.tla:203]: can you add a comment to this? It's not clear to me what the role of `r` is

**Review by heidihoward — APPROVED** [review]

File: pirateship.tla (hunk 5 above)
Comment (lemmy) [reply pirateship.tla:203]: https://github.com/heidihoward/pirateship-tla/pull/22/commits/22f68e48db65f475d88460d3eebdb271d1cb0b17

File: pirateship.tla (hunk 4 above)
Comment (lemmy) [reply pirateship.tla]: https://github.com/heidihoward/pirateship-tla/pull/22/commits/3061a0e631f75b284bcd3d76678c5b2313dab744

**Comment by heidihoward** [comment]
@lemmy Do you know why this changes causes tlc-simulate to go red?

**Comment by lemmy** [comment]
> @lemmy Do you know why this changes causes tlc-simulate to go red?

It’s failing because `ByzCommitIndex@1` remains 0. ~Surprisingly, `ByzCommitIndex@2` doesn’t persistently remain 0~.  Will investigate...

* `ByzCommitIndex@2 # 0` because `TLCpirateship!TLCInit` defines an initial state with `byzCommitIndex = [r \in R |-> 2]`. `ByzCommitIndex@2` also remains `0` if `TLCInit` is not substituted for `Init`.
* git bisect identifies commit aaf24e03d8e3d9413d56fff94de894b557941fa6 as the source of the regression.

**Pull request merged**